peps/pep-0734.rst

```diff
@@ -0,0 +1,800 @@
+
+* ``put_nowait(obj)``
+      Like ``put()`` but effectively with an immediate timeout.
+
+* ``get(timeout=None) -> object``
+      Pop the next object from the queue and return it.  Block while
+      the queue is empty.  If a positive *timeout* is provided and an
+      object hasn't been added to the queue in that many seconds
+      then raise ``interpreters.QueueEmpty``.
+
+* ``get_nowait([default]) -> object``
+      Like ``get()``, but do not block.  If the queue is not empty
+      then return the next item.  Otherwise, return *default* if
+      one was provided else raise ``interpreters.QueueEmpty``.
+
+Shareable Objects
+-----------------
+
+Both ``Interpreter.prepare___main__()`` and ``Queue`` work only with
+"shareable" objects.
+
+A "shareable" object is one which may be passed from one interpreter
+to another.  The object is not necessarily actually shared by the
+interpreters.  However, the object in the one interpreter is guaranteed
+to exactly match the corresponding object in the other interpreter.
+
+For some types (builtin singletons), the actual object is shared.
+For some, the object's underlying data is actually shared but each
+interpreter has a distinct object wrapping that data.  For all other
+shareable types, a strict copy or proxy is made such that the
+corresponding objects continue to match exactly.  In cases where
+the underlying data is complex but must be copied (e.g. ``tuple``),
+the data is serialized as efficiently as possible.
+
+Shareable objects must be specifically supported internally
+by the Python runtime.  However, there is no restriction against
+adding support for more types later.
+
+Here's the initial list of supported objects:
+
+* ``str``
+* ``bytes``
+* ``int``
+* ``float``
+* ``bool`` (``True``/``False``)
+* ``None``
+* ``tuple`` (only with shareable items)
+* ``Queue``
+* ``memoryview`` (underlying buffer actually shared)
+
+Note that the last two on the list, queues and ``memoryview``, are
+technically mutable data types, whereas the rest are not.  When any
+interpreters share mutable data there is always a risk of data races.
+Cross-interpreter safety, including thread-safety, is a fundamental
+feature of queues.
+
+However, ``memoryview`` does not have any native accommodations.
+The user is responsible for managing thread safety, whether passing
+a token back and forth through a queue to indicate safety
+(see `Synchronization`_), or by assigning sub-range exclusivity
+to individual interpreters.
+
+Regarding queues, the primary situation for sharing is where an
+interpreter is using ``prepare___main__()`` to provide the target
+interpreter with their means of further communication.
+
+Finally, a reminder: for some types the actual object is shared,
+whereas for others only the underlying data (or even a copy or proxy)
+is shared.  Regardless, the guarantee of "shareable" objects is that
+corresponding objects in different interpreters will always strictly
+match each other.
```

Comment:
I guess these are weasel words because you can't use `__eq__` and for NaNs that wouldn't even work. :-) I assume unicode normalization isn't going to be a problem.

Reply:
> I guess these are weasel words because you can't use `__eq__` and for NaNs that wouldn't even work. :-)

By "strictly match" I meant a deeper sort of equivalence than just equality.  The "shared" object would be indistinguishable from the original, whether it were the actual original object or a copy/proxy.  For oddballs like `NaN` or `inf` you'd get the right object on the other side (i.e.. `math.isnam()` and `math.isinf()`, respectively, would return true).  For mutable data, like buffers, mutation in one interpreter would be reflected in every other interpreter in which the object had been shared through a queue (or `prepare_main()`).

I'm definitely open to clearer phrasing.  In the meantime, I'll change it to "strictly equivalent".  Regardless, I'll try clarifying things with some additional text.

> I assume unicode normalization isn't going to be a problem.

Yeah, it shouldn't be a problem.

Reply:
I think being explicit about what you just told me in the comment would be good. Especially the part about mutable buffers being (or acting as) truly shared. (Though all the C code manipulating these probably will need to use atomic operations to ensure the compiler actually writes to the buffer when you expect it to.)